revscoring/extractors/api.py

```diff
@@ -13,19 +27,300 @@ def __init__(self, session, language=None):
+
+                for rev_doc in rev_docs.values():
+                    extract_caches[rev_doc['revid']][user_doc] = \
+                            user_docs.get(rev_doc.get('user'))
+
+        # Now request features one-by-one
+        for rev_id in rev_ids:
+            try:
+                values = self.extract(rev_id, features, extract_caches[rev_id])
+                yield None, values
+            except Exception as e:
+                yield e, None
+
+
+    def get_namespace_map(self):
+        logger.info("Requesting site info from the API")
+        doc = self.session.site_info.query(
+            properties={'general', 'namespaces', 'namespacealiases'}
+        )
+
+        return self.namespace_map_from_doc(doc)
+
+    def get_rev_doc_map(self, rev_ids, props={'ids', 'user', 'timestamp',
+                                              'userid', 'comment', 'content',
+                                              'flags', 'size'}):
+        logger.info("Batch requesting {0} revisions from the API" \
+                    .format(len(rev_ids)))
+        return {rd['revid']:rd
+                for rd in self.session.revisions.query(
+                    revids=rev_ids,
+                    properties=props
+                )}
+
+    def get_user_doc_map(self, user_texts, props={'blockinfo', 'implicitgroups',
+                                                  'groups', 'registration',
+                                                  'emailable', 'editcount',
+                                                  'gender'}):
+        logger.info("Batch requesting {0} users from the API" \
+                    .format(len(user_texts)))
+        return {ud['name']:ud
+                for ud in self.session.users.query(
+                    users=user_texts,
+                    properties=props
+                )}
+
+    def process_revision_doc(self, rev_id):
+        logger.info("Requesting a revision ({0}) from the API".format(rev_id))
+        props = {'ids', 'user', 'timestamp', 'userid', 'comment',
+                 'content', 'flags', 'size'}
+        try:
+            return self.session.revisions.get(rev_id=rev_id, properties=props)
+        except KeyError:
+            raise RevisionDocumentNotFound({'rev_id': rev_id})
+
+    def process_parent_revision_doc(self, revision_metadata):
+        props = {'ids', 'user', 'timestamp', 'userid', 'comment',
+                 'content', 'flags', 'size'}
+        if revision_metadata.parent_id is not None and \
+            revision_metadata.parent_id > 0:
+            rev_id = revision_metadata.parent_id
+            try:
+                logger.info("Requesting a parent revision ({0}) from the API" \
+                            .format(rev_id))
+                return self.session.revisions.get(
+                    rev_id=rev_id,
+                    properties=props
+                )
+            except KeyError:
+                return None
```

Comment:
It's common to not have a parent revision.  Datasources/features are expected to handle the lack of a parent_revision.  This is how we handled parent revisions before.